apis/supervisor/idp/v1alpha1/types_githubidentityprovider.go.tmpl

```diff
@@ -0,0 +1,219 @@
+	//
+	// Group names as presented to Kubernetes will always be prefixed by the GitHub organization name followed by a
+	// forward slash (e.g. "my-org/my-team"). GitHub organization login names can only contain alphanumeric characters
+	// or single hyphens, so the first forward slash `/` will be the separator between the organization login name and
+	// the team name or slug.
+	//
+	// If desired, an admin could configure identity transformation expressions on the Pinniped Supervisor's
+	// FederationDomain to further customize how these group names are presented to Kubernetes.
+	//
+	// +kubebuilder:default=slug
+	// +kubebuilder:validation:Enum=name;slug
+	Groups GitHubGroupNameAttribute `json:"groups"`
+}
+
+// GitHubClientSpec contains information about the GitHub client that this identity provider will use
+// for web-based login flows.
+type GitHubClientSpec struct {
+	// SecretName contains the name of a namespace-local Secret object that provides the clientID and
+	// clientSecret for an GitHub App or GitHub OAuth2 client.
+	//
+	// This secret must be of type "secrets.pinniped.dev/github-client" with keys "clientID" and "clientSecret".
+	SecretName string `json:"secretName"`
+}
+
+// GitHubIdentityProviderSpec is the spec for configuring an GitHub identity provider.
+type GitHubIdentityProviderSpec struct {
+	// GitHubApi allows configuration for GitHub Enterprise Server
+	GitHubApi GitHubAPIConfig `json:"gitHubAPI,omitempty"`
+
+	// Claims allows customization of the username and groups claims
+	//
+	// +optional
+	Claims GitHubClaims `json:"claims,omitempty"`
+
+	// AllowedOrganizations, when specified, indicates that only users with membership in at least one of the listed
+	// GitHub organizations may log in. In addition, the group membership presented to Kubernetes will only include
+	// teams within the listed GitHub organizations. Additional login rules or group filtering can optionally be
+	// provided as policy expression on any Pinniped Supervisor FederationDomain that includes this IDP.
+	//
+	// The configured GitHub App or GitHub OAuth App must be allowed to see membership in the listed organizations,
+	// otherwise Pinniped will not be aware that the user belongs to the listed organization or any teams
+	// within that organization.
+	//
+	// If no organizations are listed, you must set allowAllUsersToLogin to "Allow".
+	//
+	// +optional
+	AllowedOrganizations []string `json:"allowedOrganizations,omitempty"`
+
+	// AllowAllUsersToLogin must be set to "Allow" if allowedOrganizations is empty.
+	//
+	// This field only exists to ensure that Pinniped administrators are aware that an empty list of
+	// allowedOrganizations means all GitHub users are allowed to log in.
+	//
+	// +kubebuilder:default=Yes
+	// +kubebuilder:validation:Enum=Yes;No
+	RestrictGitHubUsersToAllowedOrganizations GitHubRestrictGitHubUsersToAllowedOrganizations `json:"restrictGitHubUsersToAllowedOrganizations"`
```

Comment:
Ref: https://github.com/vmware-tanzu/pinniped/pull/1900/files#r1532576355

For naming & enum values.

Reply:
Unfortunately the anchor is no longer valid since I rebased... could you clarify?

Reply:
I like https://github.com/vmware-tanzu/pinniped/pull/1900#discussion_r1539523783